vermouth/citation_parser.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2018 University of Groningen
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import re
+
+class BibTexDirector():
+    """
+    Lightweight parser for BibTex files. BibTex files
+    in general have an assorment of entries that
+    describe the corresponding sort of publication
+    to refere to and then a number required and optional
+    fields for the different types of entries. A field
+    for example would be Title giving the title of a
+    publication. The syntax in general looks as follows:
+
+    @<entry>{<some custom ID>, field = {<content>},
+                               field = {<cintent>}}
+
+    Alternatively the {} can be replaced by quotation
+    marks.
+
+    This parser only parses the version with {} as
+    used by google scholar. In addition we do not
+    check for missing fields or invalid fields. All
+    fields are accapted and no fields are required.
+    """
+    def __init__(self, force_field):
+        self.force_field = force_field
+        self.known_entries = ["article",
+                              "book",
+                              "booklet",
+                              "conference",
+                              "inbook",
+                              "incollection",
+                              "inproceedings",
+                              "manual",
+                              "mastersthesis",
+                              "misc",
+                              "phdthesis",
+                              "proceedings",
+                              "techreport",
+                              "unpublished"
+                             ]
+
+    @staticmethod
+    def prepare_file(lines):
+        """
+        Bibtex is not sensetive to line spacing so we join
+        the line as one string. Comment characters are not
+        allowed.
+        """
+        return " ".join(line.strip() for line in lines)
+
+    @staticmethod
+    def find_entries(citation_string):
+        """
+        Look in a string where `@` indicate the
+        begnning of a new entry and return the indices.
+
+        Parameters:
+        -----------
+        citation_string: str
+
+        Yields:
+        --------
+        idx
+            position of '@' in citation_string
+        """
+        for idx, token in enumerate(citation_string):
+            if token == "@":
+                yield idx
```

Comment:
```suggestion
        while True:
            idx = citation_string.find('@')
            if idx == -1:
                break
            yield idx
```

Reply:
@pckroon what is the difference here?

Reply:
Efficiency. `find` is much faster than enumerating over the string

Reply:
doesn't work will go with what is there

Reply:
Oops. 
```suggestion
        idx = 0
        while True:
            idx = citation_string.find('@', start=idx)
            if idx == -1:
                break
            yield idx
```
Feel free to leave it as is at this point though.